File: src/sarma_cli/engine/mcp_pool.py
"""MCP client pool — persistent MultiServerMCPClient lifecycle management."""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass
from typing import Any

from sarma_cli.engine.errors import McpConnectionError

logger = logging.getLogger(__name__)
DEFAULT_MCP_CONNECT_TIMEOUT = 20.0


@dataclass(frozen=True, slots=True)
class McpServerStatus:
    """Connection summary for one configured MCP server."""

    name: str
    connected: bool
    tool_count: int = 0
    error: str = ""


def _config_fingerprint(configs: dict[str, dict[str, Any]]) -> str:
    """Stable serialization of server configs for equality comparison."""
    try:
        return json.dumps(configs, sort_keys=True, default=str)
    except (TypeError, ValueError):
        return ""


def _connect_timeout(configs: dict[str, dict[str, Any]]) -> float:
    timeouts = [DEFAULT_MCP_CONNECT_TIMEOUT]
    for config in configs.values():
        timeout = config.get("timeout")
        if isinstance(timeout, (int, float)) and timeout > 0:
            timeouts.append(float(timeout))
    return min(timeouts)


class McpClientPool:
    """Manages persistent MCP client connections.

    Lazy-connects on first tool request, keeps clients alive for reuse,
    and provides health-check / reconnect on failure.
    """

    def __init__(self) -> None:
        self._client: Any | None = None  # MultiServerMCPClient
        self._server_configs: dict[str, dict[str, Any]] = {}
        self._config_fingerprint: str = ""
        self._tools: list[Any] = []  # list[BaseTool]
        self._connected: bool = False
        self._server_statuses: dict[str, McpServerStatus] = {}

    @property
    def is_connected(self) -> bool:
        return self._connected

    @property
    def tools(self) -> list[Any]:
        return list(self._tools)

    @property
    def server_statuses(self) -> list[McpServerStatus]:
        return list(self._server_statuses.values())

    async def connect(
        self, server_configs: dict[str, dict[str, Any]]
    ) -> list[Any]:
        """Connect (or reconnect) to MCP servers and return available tools.

        Args:
            server_configs: Dict mapping server name → connection config,
                            as produced by McpServerEntry.to_langchain_config().
        """
        fingerprint = _config_fingerprint(server_configs)
        if self._connected and fingerprint and fingerprint == self._config_fingerprint:
            return self._tools

        # Disconnect previous client if any
        await self.disconnect()

        self._server_configs = dict(server_configs)
        self._config_fingerprint = fingerprint
        self._server_statuses = {
            name: McpServerStatus(name=name, connected=False)
            for name in server_configs
        }

        if not server_configs:
            self._tools = []
            self._connected = True
            return self._tools

        try:
            from langchain_mcp_adapters.client import MultiServerMCPClient

            self._client = MultiServerMCPClient(
                server_configs, tool_name_prefix=True
            )
            self._tools = await asyncio.wait_for(
                self._client.get_tools(),
                timeout=_connect_timeout(server_configs),
            )
            self._connected = True
            self._server_statuses = {
                name: McpServerStatus(
                    name=name,
                    connected=True,
                    tool_count=self._count_server_tools(name),
                )
                for name in server_configs
            }
            logger.info(
                "MCP pool connected: %d tools from %d servers",
                len(self._tools),
                len(server_configs),
            )
            return self._tools
        except Exception as exc:
            await self.disconnect()
            self._server_configs = dict(server_configs)
            self._server_statuses = {
                name: McpServerStatus(
                    name=name,
                    connected=False,
                    error=str(exc),
                )
                for name in server_configs
            }
            logger.error("MCP pool connection failed: %s", exc)
            raise McpConnectionError(
                ", ".join(server_configs.keys()), str(exc)
            ) from exc

    async def reconnect(self) -> list[Any]:
        """Reconnect using the last known server configs."""
        if not self._server_configs:
            return []
        return await self.connect(self._server_configs)

    async def disconnect(self) -> None:
        """Cleanly close all MCP connections."""
        if self._client is not None:
            try:
                if hasattr(self._client, "close"):
                    await self._client.close()
            except Exception as exc:
                logger.warning("Error closing MCP client: %s", exc)
            finally:
                self._client = None
        self._tools = []
        self._connected = False
        self._config_fingerprint = ""
        self._server_statuses = {
            name: McpServerStatus(name=name, connected=False)
            for name in self._server_configs
        }

    def filter_tools(
        self,
        tools: list[Any],
        allowlist: set[str] | None = None,
        denylist: set[str] | None = None,
    ) -> list[Any]:
        """Apply allow/deny lists to a set of tools."""
        result = tools
        if allowlist is not None:
            result = [t for t in result if t.name in allowlist]
        if denylist is not None:
            result = [t for t in result if t.name not in denylist]
        return result

    def _count_server_tools(self, server_name: str) -> int:
        return sum(
            1
            for tool in self._tools
            if _tool_belongs_to_server(getattr(tool, "name", ""), server_name)
        )


def _tool_belongs_to_server(tool_name: str, server_name: str) -> bool:
    return (
        tool_name == server_name
        or tool_name.startswith(f"{server_name}_")
        or tool_name.startswith(f"{server_name}__")
        or tool_name.startswith(f"{server_name}.")
        or tool_name.startswith(f"{server_name}:")
    )

File: src/engine/mcpPool.ts
/** MCP client pool — persistent MultiServerMCPClient lifecycle management. */

import { MultiServerMCPClient } from "@langchain/mcp-adapters";
import type { StructuredToolInterface } from "@langchain/core/tools";
import { McpConnectionError } from "@/engine/errors";

export const DEFAULT_MCP_CONNECT_TIMEOUT = 20_000; // ms

/** Connection summary for one configured MCP server. */
export interface McpServerStatus {
  name: string;
  connected: boolean;
  toolCount: number;
  error: string;
}

type ServerConfigs = Record<string, Record<string, unknown>>;

/** Stable serialization of server configs for equality comparison. */
function configFingerprint(configs: ServerConfigs): string {
  try {
    return stableStringify(configs);
  } catch {
    return "";
  }
}

function stableStringify(value: unknown): string {
  return JSON.stringify(value, (_key, val) => {
    if (val && typeof val === "object" && !Array.isArray(val)) {
      const sorted: Record<string, unknown> = {};
      for (const k of Object.keys(val as object).sort()) {
        sorted[k] = (val as Record<string, unknown>)[k];
      }
      return sorted;
    }
    return val;
  });
}

/** Connect timeout in ms = min(default, configured server timeouts). */
function connectTimeout(configs: ServerConfigs): number {
  const timeouts = [DEFAULT_MCP_CONNECT_TIMEOUT];
  for (const config of Object.values(configs)) {
    const t = config.timeout;
    if (typeof t === "number" && t > 0) {
      // Server timeouts are expressed in seconds (parity with Python config).
      timeouts.push(t * 1000);
    }
  }
  return Math.min(...timeouts);
}

function toolBelongsToServer(toolName: string, serverName: string): boolean {
  return (
    toolName === serverName ||
    toolName.startsWith(`${serverName}_`) ||
    toolName.startsWith(`${serverName}__`) ||
    toolName.startsWith(`${serverName}.`) ||
    toolName.startsWith(`${serverName}:`)
  );
}

async function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  let timer: ReturnType<typeof setTimeout> | undefined;
  const timeout = new Promise<never>((_resolve, reject) => {
    timer = setTimeout(() => reject(new Error(`timed out after ${ms}ms`)), ms);
  });
  try {
    return await Promise.race([promise, timeout]);
  } finally {
    if (timer) clearTimeout(timer);
  }
}

/**
 * Manages persistent MCP client connections.
 *
 * Lazy-connects on first tool request, keeps clients alive for reuse, and
 * provides health-check / reconnect on failure.
 */
export class McpClientPool {
  private client: MultiServerMCPClient | null = null;
  private serverConfigs: ServerConfigs = {};
  private fingerprint = "";
  private toolList: StructuredToolInterface[] = [];
  private connected = false;
  private statuses = new Map<string, McpServerStatus>();

  get isConnected(): boolean {
    return this.connected;
  }

  get tools(): StructuredToolInterface[] {
    return [...this.toolList];
  }

  get serverStatuses(): McpServerStatus[] {
    return [...this.statuses.values()];
  }

  /**
   * Connect (or reconnect) to MCP servers and return available tools.
   *
   * @param serverConfigs map of server name → connection config, as produced
   *        by {@link McpServerDTO.toLangchainConfig}.
   */
  async connect(serverConfigs: ServerConfigs): Promise<StructuredToolInterface[]> {
    const fingerprint = configFingerprint(serverConfigs);
    if (this.connected && fingerprint && fingerprint === this.fingerprint) {
      return this.toolList;
    }

    await this.disconnect();

    this.serverConfigs = { ...serverConfigs };
    this.fingerprint = fingerprint;
    this.statuses = new Map(
      Object.keys(serverConfigs).map((name) => [
        name,
        { name, connected: false, toolCount: 0, error: "" },
      ]),
    );

    if (Object.keys(serverConfigs).length === 0) {
      this.toolList = [];
      this.connected = true;
      return this.toolList;
    }

    try {
      this.client = new MultiServerMCPClient({
        mcpServers: serverConfigs as never,
        prefixToolNameWithServerName: true,
        additionalToolNamePrefix: "",
        throwOnLoadError: true,
      });
      this.toolList = (await withTimeout(
        this.client.getTools(),
        connectTimeout(serverConfigs),
      )) as StructuredToolInterface[];
      this.connected = true;
      this.statuses = new Map(
        Object.keys(serverConfigs).map((name) => [
          name,
          { name, connected: true, toolCount: this.countServerTools(name), error: "" },
        ]),
      );
      return this.toolList;
    } catch (exc) {
      await this.disconnect();
      this.serverConfigs = { ...serverConfigs };
      const message = exc instanceof Error ? exc.message : String(exc);
      this.statuses = new Map(
        Object.keys(serverConfigs).map((name) => [
          name,
          { name, connected: false, toolCount: 0, error: message },
        ]),
      );
      throw new McpConnectionError(Object.keys(serverConfigs).join(", "), message);
    }
  }

  /** Reconnect using the last known server configs. */
  async reconnect(): Promise<StructuredToolInterface[]> {
    if (Object.keys(this.serverConfigs).length === 0) return [];
    return this.connect(this.serverConfigs);
  }

  /** Cleanly close all MCP connections. */
  async disconnect(): Promise<void> {
    if (this.client !== null) {
      try {
        await this.client.close();
      } catch {
        /* best-effort close */
      } finally {
        this.client = null;
      }
    }
    this.toolList = [];
    this.connected = false;
    this.fingerprint = "";
    this.statuses = new Map(
      Object.keys(this.serverConfigs).map((name) => [
        name,
        { name, connected: false, toolCount: 0, error: "" },
      ]),
    );
  }

  /** Apply allow/deny lists to a set of tools. */
  filterTools(
    tools: StructuredToolInterface[],
    allowlist: Set<string> | null = null,
    denylist: Set<string> | null = null,
  ): StructuredToolInterface[] {
    let result = tools;
    if (allowlist !== null) {
      result = result.filter((t) => allowlist.has(t.name));
    }
    if (denylist !== null) {
      result = result.filter((t) => !denylist.has(t.name));
    }
    return result;
  }

  private countServerTools(serverName: string): number {
    return this.toolList.filter((t) => toolBelongsToServer(t.name ?? "", serverName)).length;
  }
}
